106-auto-rebalance-on-imbalanced-clusters.md

```diff
@@ -0,0 +1,518 @@
+metadata:
+  name: my-cluster-auto-rebalancing-imbalance-<anomalyID>
+  finalizers:
+    - strimzi.io/auto-rebalancing
+spec:
+  mode: full
+  goals:
+    - CpuCapacityGoal
+    - NetworkInboundCapacityGoal
+    - DiskCapacityGoal
+    - RackAwareGoal
+    - MinTopicLeadersPerBrokerGoal
+    - NetworkOutboundCapacityGoal
+    - ReplicaCapacityGoal
+  skipHardGoalCheck: true
+  # ... other rebalancing related configuration
+```
+
+The operator also sets a finalizer, named `strimzi.io/auto-rebalancing`, on the generated `KafkaRebalance` custom resource.
+This is needed to avoid the user, or any other tooling, deleting the resource while the auto-rebalancing is still running.
+The finalizer is removed when Cruise Control indicates that the partition reassignment (rebalance) process has finished allowing the generated `KafkaRebalance` custom resource to be deleted by the operator itself.
+In case the rebalance finishes with an error, the error message will be propagated to the Kafka custom resource just like we do for the `remove-broker` and `add-broker` endpoint and the generated `KafkaRebalance` will be deleted.
+
+#### StrimziCruiseControlNotifier
+
+Cruise Control provides the `AnomalyNotifier` interface, which has multiple abstract methods on what to do if certain anomalies are detected.
+Some of those methods are: `onGoalViolation()`, `onBrokerFailure()`, `onDiskFailure`, `alert()`.
+The `StrimziCruiseControlNotifier` will implement the `AnomalyNotifier` interface.
+Every detected anomaly is placed in a priority queue and each of the anomalies have a priority value assigned to them.
+Since we are working with only goal related violations, every anomaly will have same priority.
+This means that detected anomalies will be fixed on a first come, first served basis.
+In situations where the anomaly is unfixable, due to issues like a lack of physical hardware (e.g. insufficient number of racks to satisfy rack awareness, insufficient number of brokers to satisfy `ReplicaCapacityGoal`, or insufficient number of resources to satisfy resource capacity goals), then they will be ignored and the user will be alerted via a condition in the Kafka CR status.
+The `StrimziCruiseControlNotifier` will be kept in a separate repository and will override all the methods declared by the `AnomalyNotifier` interface.
+We will provide a concrete implementations of the `AnomalyNotifier.alert` method which will alert the operator whenever an anomaly is detected by the operator.
+Upon detection of an anomaly, the notifier would create (if it doesn't already exist) or update a ConfigMap with the name `<cluster-name>-goal-violation-map`.
+This ConfigMap will contain all the detected anomalies for a given Kafka cluster.
+
+The ConfigMap will look like this:
+
+```yaml
+kind: ConfigMap
+apiVersion: v1
+metadata:
+  name: my-cluster-goal-violation-map
+data:
+  fixableAnomalies: |
+    - <anomaly-id-1>:
+        - <fixable-goal-1>
+        - <fixable-goal-2>
```

Comment:
Just to be sure I got it. Whenever Cruise Control an anomaly, it can bring more goal violations with it? Any example?

Reply:
Never mind, I found the answer by looking at the code. It's clearly possible by looking at here https://github.com/linkedin/cruise-control/blob/main/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/AnomalyDetails.java#L69